File: web/js/components/reminders.js
async function getReminders(folder) {
  let formAction = `sendTemplate({}, 'form', 'reminders', 'formHolder')`;
  let reminders = await eel.get_reminders(folder)();
  let html = await loadTemplate({ 'formAction': formAction, 'title': 'reminders', 'icon': 'bi bi-calendar-plus' }, 'sectionHeader', 'folders');
  html += await eel.render_template_list(reminders, 'reminders', 'items')();
  document.getElementById('reminders').innerHTML = html;
}

async function deleteReminder(id) {
  await eel.delete_reminder(id)();
  getReminders(getCookie('folder'));
}

async function createReminder() {
  let name = document.getElementById('reminderName').value;
  let folder_id = getCookie('folder');
  let frequency = document.getElementById('frequency').value;
  let time = null; // your format might vary
  let week = null; // for example, Wednesday
  let month = null;
  let month_of_year = null;
  let end_datetime = null;

  switch (frequency) {
    case 'none':
      end_datetime = document.getElementById('end_datetime').value;
      break;
    case 'daily':
      time = document.getElementById('time').value;
      break;
    case 'weekly':
      week = document.getElementById('day_of_week').value;
      time = document.getElementById('time').value;
      break;
    case 'monthly':
      month = document.getElementById('day_of_month').value;
      time = document.getElementById('time').value;
      break;
    case 'yearly':
      month_of_year = document.getElementById('month_of_year').value;
      month = document.getElementById('day_of_month').value;
      time = document.getElementById('time').value;
      break;
  }

  await eel.create_reminder(name, folder_id, frequency, time, week, month, month_of_year, end_datetime, false)();
  await sendTemplate({}, 'void', 'home', 'formHolder');
  await getReminders(getCookie('folder'));
}

async function loadReminder(id) {
  let reminder = await eel.get_reminder(id)();
  sendTemplate(reminder, 'page', 'reminders', 'popHolder');
}

eel.expose(trigger_ui);
function trigger_ui(name, id, html) {
  getReminders(getCookie('folder'));
  if (document.getElementById('popHolder').innerHTML == "") {
    sendTemplate({'name': name, 'id': id, 'html': html}, 'popup', 'reminders', 'popHolder');
  } else {
    document.getElementById('popHolder').innerHTML += loadTemplate({'html': html, 'name': name, 'id': id}, 'popup', 'reminders');
  }
}

function updateRemindForm(frequency) {
  let dayOfWeekDiv = document.getElementById('day_of_week_div');
  let dayOfMonthDiv = document.getElementById('day_of_month_div');
  let monthOfYearDiv = document.getElementById('month_of_year_div');
  let endDatetimeDiv = document.getElementById('end_datetime_div');
  let timeDiv = document.getElementById('time_div');

  // Initially hide all optional fields
  timeDiv.style.display = 'flex';
  dayOfWeekDiv.style.display = 'none';
  dayOfMonthDiv.style.display = 'none';
  monthOfYearDiv.style.display = 'none';
  endDatetimeDiv.style.display = 'none';

  // Display the appropriate fields based on the selected frequency
  switch (frequency) {
    case 'none':
      endDatetimeDiv.style.display = 'flex';
      timeDiv.style.display = 'none';
      break;
    case 'weekly':
      dayOfWeekDiv.style.display = 'flex';
      break;
    case 'monthly':
      dayOfMonthDiv.style.display = 'flex';
      break;
    case 'yearly':
      monthOfYearDiv.style.display = 'flex';
      dayOfMonthDiv.style.display = 'flex';
      break;
  }
}

File: api/reminders.py
import sys
sys.path.append("..")
import eel
from models import Reminder
from datetime import datetime
from api.attachments import delete_all_attachments, delete_attachment

# reminders api

def reminder_freq_set(reminder_dict, reminder):
  if reminder.frequency == 'none' and reminder.end_datetime is not None:
    reminder_dict['end_datetime'] = reminder.end_datetime.strftime("%Y-%m-%d %I:%M%p")

  elif reminder.frequency == 'daily' and reminder.time_of_day is not None:
    reminder_dict['time'] = reminder.time_of_day.strftime("%I:%M%p")

  elif reminder.frequency == 'weekly' and reminder.time_of_day is not None and reminder.day_of_week is not None:
    reminder_dict['time'] = reminder.time_of_day.strftime("%I:%M%p")
    reminder_dict['week'] = reminder.day_of_week

  elif reminder.frequency == 'monthly' and reminder.time_of_day is not None and reminder.day_of_month is not None:
    reminder_dict['time'] = reminder.time_of_day.strftime("%I:%M%p")
    reminder_dict['month'] = reminder.day_of_month

  elif reminder.frequency == 'yearly' and reminder.time_of_day is not None and reminder.day_of_month is not None and reminder.month_of_year is not None:
    reminder_dict['time'] = reminder.time_of_day.strftime("%I:%M%p")
    reminder_dict['month'] = reminder.day_of_month
    reminder_dict['month_of_year'] = reminder.month_of_year


@eel.expose
def get_reminders(folder_id):
    reminders = Reminder.select().where(Reminder.folder_id == folder_id)
    current_time = datetime.now()
    reminders_list = []

    for reminder in reminders:
        reminder_dict = {'id': reminder.id, 'name': reminder.name, 'frequency': reminder.frequency, 'deleted': reminder.deleted, 'current_time': current_time.strftime("%Y-%m-%d %I:%M%p").lower()}
        reminder_freq_set(reminder_dict, reminder)
        reminders_list.append(reminder_dict)

    return reminders_list

@eel.expose
def get_reminder(id):
    reminder = Reminder.get(Reminder.id == id)
    current_time = datetime.now()
    reminder_dict = {'id': reminder.id, 'name': reminder.name, 'frequency': reminder.frequency, 'deleted': reminder.deleted, 'current_time': current_time.strftime("%Y-%m-%d %I:%M%p").lower()}
        
    reminder_freq_set(reminder_dict, reminder)

    return reminder_dict
    

@eel.expose
def delete_reminder(id):
    reminder = Reminder.get(Reminder.id == id)
    reminder.deleted = True
    delete_all_attachments(id, 'reminder')
    reminder.save()


def get_all_reminders():
    return Reminder.select()


@eel.expose
def create_reminder(
        name, folder_id,
        frequency,
        time,
        week=None,
        month=None,
        month_of_year=None,
        end_datetime=None,
        deleted=None
):
    if end_datetime:
        end_datetime = datetime.strptime(end_datetime, "%Y-%m-%dT%H:%M")

    if time:
        time = datetime.strptime(time, '%H:%M').time()

    reminder = Reminder.create(
        name=name,
        folder=folder_id,
        frequency=frequency,
        time_of_day=time,
        day_of_week=week,
        day_of_month=month,
        month_of_year=month_of_year,
        end_datetime=end_datetime,
        deleted=deleted
    )

    from scheduler import schedule_reminder, run_reminder
    schedule_reminder(reminder)


